src/ImageSharp/Formats/Ico/Common/IcoLikeDecoderCore.cs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using SixLabors.ImageSharp;
+using SixLabors.ImageSharp.IO;
+using SixLabors.ImageSharp.Metadata;
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Formats.Ico.Common;
+
+internal sealed class IcoLikeDecoderCore : IImageDecoderInternals
+{
+    private IcoDir icoDir;
+    private IcoDirEntry[] icoDirEntries = Array.Empty<IcoDirEntry>();
+
+    private readonly WeakReference<Png.PngDecoderCore?> pngDecoder = new(default);
+    private readonly WeakReference<Bmp.BmpDecoderCore?> bmpDecoder = new(default);
+
+    public IcoLikeDecoderCore(DecoderOptions options) => this.Options = options;
+
+    public DecoderOptions Options { get; }
+
+    public Size Dimensions { get; private set; }
+
+    public Image<TPixel> Decode<TPixel>(BufferedReadStream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        ReadHeader(stream, out this.icoDir, out this.icoDirEntries);
+        this.Dimensions = new(
+             this.icoDirEntries.Max(i => i.Width),
+             this.icoDirEntries.Max(i => i.Height));
+
+        Span<byte> flag = stackalloc byte[Png.PngConstants.HeaderBytes.Length];
+
+        List<(Image<TPixel> Image, bool IsPng, int Index)> frames = new(this.icoDirEntries.Length);
+        for (int i = 0; i < this.icoDirEntries.Length; i++)
+        {
+            _ = stream.Seek(this.icoDirEntries[i].ImageOffset, SeekOrigin.Begin);
+
+            int size = stream.Read(flag);
+            _ = stream.Seek(-size, SeekOrigin.Current);
+
+            bool isPng = flag.SequenceEqual(Png.PngConstants.HeaderBytes);
+
+            Image<TPixel> img = this.GetDecoder(isPng).Decode<TPixel>(stream, cancellationToken);
+            IcoAssert.NotSquare(img.Size);
+            frames.Add((img, isPng, i));
```

Comment:
We should be decoding each image in full here so we can handle early EOF. 

I would add an internal method to `ImageFrameCollection<T>` called `ExtractFrameUnsafe()` (_unsafe because it can leave an image containing a single frame in an unusable state_) which allows removing a frame from an image at a given index. You can then dispose of the parent safely. 

A major advantage to this other than the lack of copying (_and why I keep pushing you to use the original metadata_)  is that you already have the bmp or png metadata in the image frame. You only now need to add your ico/cur metadata